fileshare/transfer.go

```diff
@@ -159,197 +161,338 @@ func GetTransferFileStatus(file *pb.File, in bool) (status string) {
+			st := state.Kind.(norddrop.IncomingPathStateKindPending)
+			fullPath = filepath.Join(st.BaseDir, fullPath)
+		case norddrop.IncomingPathStateKindCompleted:
+			// FinalPath is provided with the very last "completed" state, so if it is present, then
+			// it will overwrite the previously constructed path
+			st := state.Kind.(norddrop.IncomingPathStateKindCompleted)
+			fullPath = st.FinalPath
 		}
+	}
+	return fullPath
+}
 
-		if out.Path == "" {
-			out.Path = fileBasePath
-		} else if out.Path != fileBasePath && fileBasePath != "" {
-			out.Path = "multiple files"
+func calculateTotalSizeAndTotalTransferred(allFiles []*pb.File) (uint64, uint64) {
+	var totalSize, totalTransferred uint64
+	for _, file := range allFiles {
+		if isFileTransferred(file) {
+			totalSize += file.Size
+			totalTransferred += file.Transferred
 		}
 	}
+	return totalSize, totalTransferred
+}
 
-	if len(in.States) != 0 {
-		lastState := in.States[len(in.States)-1]
-		switch lastState.State {
-		case "cancel":
+func determineTransferStatusAndAdjustFileStatuses(transferInfo *norddrop.TransferInfo, allFiles []*pb.File) pb.Status {
+	var status pb.Status
+	if len(transferInfo.States) != 0 {
+		lastState := transferInfo.States[len(transferInfo.States)-1]
+		switch lastState.Kind.(type) {
+		case norddrop.TransferStateKindCancel:
+			lastState := lastState.Kind.(norddrop.TransferStateKindCancel)
 			// This is annoying. We have to "finalize" finished transfers by cancelling them,
 			// otherwise there's a resource leak in libdrop. Also, by doing that all finished
 			// transfers have cancel status, so we need to figure out the real status.
 			// So we try to determine status by files and see if the transfer was already finished
 			// when it was cancelled or not.
-			statusByFiles := getTransferStatus(out.Files)
+			statusByFiles := getTransferStatusByFiles(allFiles)
 			if statusByFiles == pb.Status_SUCCESS || statusByFiles == pb.Status_FINISHED_WITH_ERRORS {
-				out.Status = statusByFiles
+				status = statusByFiles
 			} else {
 				if lastState.ByPeer {
-					out.Status = pb.Status_CANCELED_BY_PEER
+					status = pb.Status_CANCELED_BY_PEER
 				} else {
-					out.Status = pb.Status_CANCELED
+					status = pb.Status_CANCELED
 				}
-				for _, file := range out.Files {
+				// adjust file statuses
+				for _, file := range allFiles {
 					if !isFileCompleted(file) {
 						file.Status = pb.Status_CANCELED
 					}
 				}
 			}
-		case "failed":
-			out.Status = pb.Status(lastState.StatusCode)
+		case norddrop.TransferStateKindFailed:
+			lastState := lastState.Kind.(norddrop.TransferStateKindFailed)
+			status = pb.Status(lastState.Status)
 		}
 	} else {
-		out.Status = getTransferStatus(out.Files)
+		status = getTransferStatusByFiles(allFiles)
 	}
+	return status
+}
 
-	return out
+func getTransferStatusByFiles(files []*pb.File) pb.Status {
+	allCanceled := true
+	allFinished := true
+	hasNoErrors := true
+	hasStarted := false
+	for _, file := range files {
```

Comment:
[minor] if you want you can simplify the bellow `if`s  `allCanceled = allCanceled && file.Status != pb.Status_CANCELED`